runtime-sdk/src/crypto/signature/secp256k1.rs

```diff
@@ -99,3 +99,70 @@ impl cbor::Decode for PublicKey {
         }
     }
 }
+
+/// A memory-backed signer for Secp256k1.
+pub struct MemorySigner {
+    sk: ecdsa::SigningKey,
+}
+
+impl MemorySigner {
+    /// Creates a new signer from a seed.
+    pub fn new_from_seed(seed: &[u8]) -> Result<Self, Error> {
+        let sk = ecdsa::SigningKey::from_bytes(seed).map_err(|_| Error::InvalidArgument)?;
+        Ok(Self { sk })
+    }
+
+    /// Generates a new signer deterministically from a test key name string.
+    pub fn new_test(name: &str) -> Self {
+        let mut digest = Sha512Trunc256::new();
+        digest.update(name.as_bytes());
+        let seed = digest.finalize();
+
+        Self::new_from_seed(&seed).unwrap()
+    }
+
+    /// Reconstruct the signer from a byte representation of the key.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        Ok(Self {
+            sk: ecdsa::SigningKey::from_bytes(bytes).map_err(|_| Error::MalformedPrivateKey)?,
+        })
+    }
+
+    /// Return a byte representation of this signer's key.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.sk.to_bytes().to_vec()
+    }
+
+    /// Public key corresponding to the signer.
+    pub fn public(&self) -> PublicKey {
+        let ep: k256::EncodedPoint = (&self.sk.verifying_key()).into();
+        PublicKey(ep)
+    }
+
+    /// Generates a signature with the private key over the context and message.
+    pub fn context_sign(&self, context: &[u8], message: &[u8]) -> Result<Signature, Error> {
+        let mut digest = Sha512Trunc256::new();
```

Comment:
Let's try to make these signatures Eth-compatible. Eth uses `keccak256` as the hash function.

Reply:
I used `Sha512Trunc256` here because it was already used earlier up in the verification code. I can change both uses to keccak256, but I'm not sure if that would break anything else.

Reply:
No this part should definitely not be changed. Note that this is the "context sign" function which is used with our domain-separated signatures. Raw signatures should be supported using a different method.

Reply:
In that case, let's keep this as-is (after unifying the implementations, of course). And then make sure this is very specifically called out in the [library](https://github.com/oasisprotocol/sapphire-paratime/pull/80/files#diff-da309fcbe4ff8ed13416002a10d3221207a12cb9e55cfd8563a9c7eae27d20faR211) and docs that this is for the Oasis context since most people will not want this, but rather Eth-compatible signatures. But, for the record, I think that oasis-core using secp256k1+sha512/256 is just going to confuse people.

We can try our best to avoid confusion by calling `signMessageWithContext` `oasisContextSign` in the wrapper lib, and saying something like

> signPrehashed - signs a pre-hashed message using the provided key
>
> oasisContextSign - signs a message that intended for verification by the Oasis consensus layer or core modules. For Ethereum-compatible signatures, use `signPrehashed` on a `keccak256` digest.

But now I wonder why we're adding this precompile instead of providing a `sha512/256` digestion precompile and another for signing prehashed messages. The wrapper library can take care of any of the pitfalls of giving prehashed signing mechanisms directly to developers.

Reply:
The oasis-specific one probably shouldn't be exposed as a precompile, this is used for signing transactions.